Policies, Groups, Password

firsti opened the pull request:
This PR adds the functionality to update and reset user passwords, add,  assign and delete groups and modify (set and delete) policies and single locks.

The schemas to add entities are fetched from IDM.
Tabs are generated dynamically by using the configuration for the UI. Furthermore, properties of attributes and tabs in the UI configuration file are also used (display names/hidden fields/editable fields etc.).

This PR uses agile-security [v3.3.0](https://github.com/Agile-IoT/agile-security/tree/v3.3.0)

firsti committed:
Add password fields, create new entities, create and assign groups

firsti committed:
Add react-jsonschema-form to package.json

firsti committed:
Replace tabs with 2 spaces

firsti committed:
Remove custom sdk path

firsti committed:
Remove spaces

firsti committed:
Rename newToken

firsti committed:
Read, list and delete policies/locks

firsti committed:
Add locks

firsti committed:
Add and delete entity

firsti committed:
Remove custom set token

firsti committed:
Delete block

firsti committed:
Add block

firsti committed:
Remove console.log

firsti committed:
Remove inline edit function

firsti committed:
Merge branch 'groups' into policies

# Conflicts:
#	src/actions/index.js

firsti committed:
Convert intends to spaces

firsti committed:
ESLint recommendations

firsti committed:
Remove div to align buttons

firsti committed:
Add empty read/write lock

firsti committed:
Hide 'not' fields

firsti committed:
Make sure generic entites have ID field when adding them

firsti committed:
isOwner Lock

firsti committed:
Add Lock multiple attributes

firsti committed:
Add tooltip to form

firsti committed:
Add names to locks

firsti committed:
Replace all tabs with spaces

firsti committed:
Remove console.log

firsti committed:
Remove custom SDK location

firsti committed:
Reset view when opening AddLock.js

firsti committed:
Merge branch 'master' into policies

# Conflicts:
#	package.json

craigmulligan commented on src/actions/index.js:
```diff
@@ -254,17 +268,31 @@ export const userFetch = (id) => {
+    dispatch(loading(true))
+    agile.idm.user.resetPassword(username, authType, newPassword)
+      .then(entity => {
+        // dispatch(action('PASSWORD', entity));
```
nit: rm debug code.

craigmulligan commented on src/actions/index.js:
```diff
@@ -321,6 +361,9 @@ export const canExecuteActions = (id, type, attribute_names, actions) => {
     });
 
     attribute_names.push('actions.self'); //Add root object
+    if(attribute_names.indexOf('password') === -1 && type === 'user') {
```
nit: space after `if`. Also you can use lodash/native .includes over `.indexOf` find it easier to read at a glance.

craigmulligan commented on src/actions/index.js:
```diff
@@ -345,13 +388,78 @@ export const canExecuteActions = (id, type, attribute_names, actions) => {
 }
 
-// fetch all users
+export const currentTab = (type) => {
```
What is this used for? I'd imagine the component lib keep track of this state and have callbacks we can use?

craigmulligan commented on src/actions/index.js:
```diff
@@ -345,13 +388,78 @@ export const canExecuteActions = (id, type, attribute_names, actions) => {
-    agile.idm.entity.getByType(type)
-      .then(entities => {
-        dispatch(action('ENTITIES', entities));
+    switch(type) {
```
nit: I'd just use an `if` over a switch for a single condition.

craigmulligan commented on src/actions/index.js:
```diff
@@ -375,12 +483,69 @@ export const usersDelete = (userName, auth_type) => {
+  }
+}
+
+export const entityDeleteByType = (entity, type) => {
```
Do you not need to dispatch events here? How do we communicate a successful delete?

firsti replied:
No, since this is just a function that is used for different types of entities and it delegates to other functions, which are actually doing the delete. The other functions are dispatching the event.

craigmulligan commented on src/actions/index.js:
```diff
@@ -389,6 +554,127 @@ export const usersCreate = (user, type) => {
+      ]
+    },
+    isOwner: {
+      scopes: ["/client", "/device", "/gateway"],
```
Nit: replace `"` with `'`

craigmulligan commented on src/components/Form.js:
```diff
@@ -0,0 +1,81 @@
+  return formNames.map((formName, i) => {
+      if (forms[formName]) {
+        if (forms[formName].args) {
+          return (<div key={formName + '_' + i}
```
nit: use string interpolation `${formName}_${i}`

craigmulligan commented on src/components/LockItem.js:
```diff
@@ -0,0 +1,94 @@
+    return (<ListItem key={`arg_${i}`}>
+        {i}:  {'' + arg}
+      </ListItem>
+    );/*{(
```
Let's delete the comments

craigmulligan commented on src/components/LockItem.js:
```diff
@@ -0,0 +1,94 @@
+} from 'material-ui/Card';
+import {List, ListItem} from 'material-ui/List';
+
+const getArrayArgsListItems = (args, lock) => {
```
doesn't look like `lock` used in this function

craigmulligan commented on src/components/LockItem.js:
```diff
@@ -0,0 +1,94 @@
+        }
+        </ListItem>
+      );
+      /*<InlineEdit activeClassName='editing'
```
rm comments

craigmulligan commented on src/components/LockItem.js:
```diff
@@ -0,0 +1,94 @@
+
+const getArgsListItems = (args, lock) => {
+  let itemList = [];
+  for (var arg in args) {
```
Think a reduce maybe more concise here.

craigmulligan commented on src/components/LockItem.js:
```diff
@@ -0,0 +1,94 @@
+  return (<List>{itemList}</List>);
+}
+
+const getLocksListItems = (locks, op) => {
```
I think a small summary above each function in the component maybe do good here.

craigmulligan commented on src/components/SecurityItem.js:
```diff
@@ -96,6 +100,7 @@ const renderAttributes = (id, type, attributes, addAttributeField, dataChanged,
 }
 
 const SecurityItem = (props) => {
+  ui = props.ui;
```
nit: you can deconstruct like `const { ui } = props`

craigmulligan commented on src/containers/AddLock.js:
```diff
@@ -0,0 +1,109 @@
+          let locks = [];
+          while(event.target[i]) {
+            const lockInfo = event.target[i].name.split('_');
+            if(lockInfo[0] && lockInfo[0] !== '') {
```
could you name `lockInfo[0]` and `lockInfo[1]` so we know what they represent.

craigmulligan commented on src/containers/AddLock.js:
```diff
@@ -0,0 +1,109 @@
+        onSubmit={event => {
+          let i = 0;
+          let locks = [];
+          while(event.target[i]) {
```
What is the block of code doing exactly? Let's try make it more readable.

firsti committed:
address comments provided during the code review

firsti commented:
Hi @craig-mulligan,

thanks for the review. I have addressed your comments, e.g. improved the code or added some comments to make it more readable. Can you have a look again?

Thanks in advance!

Exulansis committed:
Refactoring in order to have a more consistent user interface.

Exulansis committed:
Merge branch 'policies' of github.com:Agile-IoT/agile-ui into policies

Exulansis committed:
Merge branch 'master' into policies

Exulansis committed:
WIP updating to subscriptions

Exulansis committed:
Merge branch 'master' into policies

Exulansis committed:
Cleaning up and synchronizing the graphs to incoming data.

Exulansis committed:
Minor refactoring to components, and final integration of UI elements for user and group management

Change-Type: major

Exulansis reviewed (approved):
The main functionality seems to be working as intended. There are further UI improvements that need to be made, but it will be more productive if those are split into smaller, atomic issues we can tackle easier.
